substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+}
+
+pub struct CallReceipt {
+	pub return_data: Vec<u8>,
+}
+
+pub struct ExecutionContext<'a, T: Trait + 'a> {
+	// typically should be dest
+	pub self_account: T::AccountId,
+	pub overlay: OverlayAccountDb<'a, T>,
+	pub depth: usize,
+}
+
+impl<'a, T: Trait> ExecutionContext<'a, T> {
+	/// Make a call to the specified address.
+	pub fn call(
+		&mut self,
+		caller: T::AccountId,
+		dest: T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		_data: Vec<u8>,
+	) -> Result<CallReceipt, &'static str> {
+		let dest_code = <CodeOf<T>>::get(&dest);
+
+		// TODO: check the new depth
+
+		let call_base_fee = <Module<T>>::call_base_fee();
+		if gas_meter.charge(call_base_fee).is_out_of_gas() {
+			return Err("not enough gas to pay base call fee");
+		}
+
+		let (exec_result, change_set) = {
+			let mut overlay = OverlayAccountDb::new(&self.overlay);
+
+			if value > T::Balance::zero() {
+				transfer(
+					gas_meter,
+					false,
+					&self.self_account,
+					&dest,
+					value,
+					&mut overlay,
+				)?;
+			}
+
+			let mut nested = ExecutionContext {
+				overlay: overlay,
+				self_account: dest.clone(),
+				depth: self.depth + 1,
+			};
+			let exec_result = if !dest_code.is_empty() {
+				vm::execute(
+					&dest_code,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while call")?
+			} else {
+				// that was a plain transfer
+				vm::ExecutionResult {
+					return_data: Vec::new(),
+				}
+			};
+
+			(exec_result, nested.overlay.into_change_set())
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CallReceipt {
+			return_data: exec_result.return_data,
+		})
+	}
+
+	pub fn create(
+		&mut self,
+		caller: T::AccountId,
+		endowment: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		ctor: &[u8],
+		_data: &[u8],
+	) -> Result<CreateReceipt<T>, &'static str> {
+		let create_base_fee = <Module<T>>::create_base_fee();
+		if gas_meter.charge(create_base_fee).is_out_of_gas() {
+			return Err("not enough gas to pay base create fee");
+		}
+
+		let dest = T::DetermineContractAddress::contract_address_for(ctor, &self.self_account);
+		if <CodeOf<T>>::exists(&dest) {
+			// TODO: Is it enough?
+			return Err("contract already exists");
+		}
+
+		let change_set = {
+			let mut overlay = OverlayAccountDb::new(&self.overlay);
+
+			if endowment > T::Balance::zero() {
+				transfer(
+					gas_meter,
+					true,
+					&self.self_account,
+					&dest,
+					endowment,
+					&mut overlay,
+				)?;
+			}
+
+			let mut nested = ExecutionContext {
+				overlay: overlay,
+				self_account: dest.clone(),
+				depth: self.depth + 1,
+			};
+			let exec_result = {
+				vm::execute(
+					ctor,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while create")?
+			};
+
+			nested.overlay.set_code(&dest, exec_result.return_data().to_vec());
```

Comment:
In `runtime/consensus/src/lib.rs:89`, we can change `set_code` to take only `&[u8]`. This will avoid the `to_vec` copy here. Or a temporary solution before we properly fix quadratic attacks -- use `exec_result.return_data` directly because we don't need `exec_result` after this point.